From f3ad3e3b8bd0e36f8df239d1c4077bf8eb8b7003 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Thu, 12 Jun 2025 18:14:29 +0200
Subject: [PATCH 195/985] net: fec: prevent IRQ threading if oob support is
 built in

We won't cope with threaded irqs if oob I/O might be enabled for a
device. Most of our interrupt-triggered work already happens out of
IRQ context anyway (NAPI softirq, workqueues), so in practice, the
cost is marginal, and at any rate, won't impact the oob mode.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 drivers/net/ethernet/freescale/fec_main.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/freescale/fec_main.c b/drivers/net/ethernet/freescale/fec_main.c
index 28dd484a2..ac5a173df 100644
--- a/drivers/net/ethernet/freescale/fec_main.c
+++ b/drivers/net/ethernet/freescale/fec_main.c
@@ -312,6 +312,14 @@ static int mii_cnt;
 
 #ifdef CONFIG_FEC_OOB
 
+/*
+ * We won't cope with threaded irqs if oob I/O might be enabled for a
+ * device. Most of our interrupt-triggered work already happens out of
+ * IRQ context anyway (NAPI softirq, workqueues), so in practice, the
+ * cost is marginal, and at any rate, won't impact the oob mode.
+ */
+#define FEC_IRQ_FLAGS  IRQF_NO_THREAD
+
 static int fec_enet_get_irq_cnt(struct platform_device *pdev);
 
 static int fec_enable_oob(struct net_device *ndev)
@@ -352,6 +360,10 @@ static void fec_disable_oob(struct net_device *ndev)
 	napi_enable(&fep->napi);
 }
 
+#else  /* !CONFIG_FEC_OOB */
+
+#define FEC_IRQ_FLAGS  0
+
 #endif	/* !CONFIG_FEC_OOB */
 
 static struct bufdesc *fec_enet_get_nextdesc(struct bufdesc *bdp,
@@ -4684,7 +4696,7 @@ fec_probe(struct platform_device *pdev)
 			goto failed_irq;
 		}
 		ret = devm_request_irq(&pdev->dev, irq, fec_enet_interrupt,
-				       0, pdev->name, ndev);
+				       FEC_IRQ_FLAGS, pdev->name, ndev);
 		if (ret)
 			goto failed_irq;
 
-- 
2.30.2

